### DISCUSSION TITLE
Unexpected behaviour Paredit: Expand Selection on comment

### DISCUSSION THREAD

**Discussion by hjer** [BetterThanTomorrow/calva#2125] [General]
When I'm using the expand selection (ctrl + w) on something like:
```
#_[:foo
      [:bar]]
```
(and my cursor is located after `#_`)

It selects `#_` and not 
```
[:foo
      [:bar]]
```
Which I thought it would. Don't know if this is intentional or not.

I expected it to work as this: if i stand in between the "" in the example below - it selects `bar` - but if i'm outside it selects `"bar"` and if I'm inside the function on `foo` it selects `foo` and if I'm outside it selects `(foo "bar")`

```
(foo "bar")
```

**Comment by PEZ** [comment]
Hi! Thanks for reporting.

There are two issues here.

1. Selection of ignored forms (the `#_`) issue.
2. General behaviour of the expand selection command.

## The Current Form

Let me start with the General case. When there is no selection, **Expand selection** will select the _Current Form_. This is an important concept in Calva, a lot of commands depend on it. A form here means about the same as an s-expression. It's either a ”solid” thing, like a symbol or a keyword or a literal number, or a list of some type (`()`, `[]`, `{}`, or `#{}`). 

* With the cursor in a ”solid”, that ”solid” is considered the current form.
* With the cursor somewhere else Calva will prioritize forms like this:
  1. Is it directly adjacent to the cursor?
  1. Is it ”near” (only separated by whitespace) the cursor?
     1. Is it on the same line?
        1. Is it before the cursor?
     1. Is it on a separate line?
        1. Is it on the previous line?

If we consider the cases for things on the same line, using your example, and use `|` to represent the cursor, we have these scenarios:

| Text and cursor | Current form    |
|-----------------|----------------|
| `(foo "ba\|r")`  | `bar`          |
| `(foo "bar\|")`  | `bar`          |
| `(foo "\|bar")`  | `bar`          |
| `(foo "bar"\|)`  | `"bar"`        |
| `(foo \|\ "bar")`  | `"bar"`      |
| `(foo\|\ "bar")`  | `foo`          |
| `(\|\ foo "bar")`  | `foo`        |
| `(foo \| "bar")`  | `foo`         |
| `(foo "bar")\|`  | `(foo "bar")` |
| ` \| (foo "bar")` | `(foo "bar")` |
| `(foo "bar") \| baz` | `(foo "bar")` |
| `(foo "bar") \|baz` | `baz`        |
| `baz \|(foo "bar")` | `(foo "bar")` |
| `baz \| (foo "bar")` | `baz`        |

Here are two places where we have described the _Current Form_ concept in different ways:

* https://calva.io/eval-tips/#current-form
* https://www.youtube.com/watch?v=8ygw7LLLU1w

## The Ignore Reader Tag (`#_`)

So far the general case and what you can expect. As for the ignored form, it could probably work a bit nicer. Currently, for the **Expand Selection** case, Calva treats `#_` as a form. So we get this behaviour:

| Text and cursor | Current form    |
|-----------------|----------------|
| `#_\|foo`  | `#_`          |
| `#_foo\|`  | `foo`          |
| `#_ \|foo`  | `foo`          |
| `#_\| foo`  | `#_`          |

I'm not sure what would be least surprising and most useful, tbh. Maybe this:

| Text and cursor | Current form    |
|-----------------|----------------|
| `#_\|foo`  | `#_foo`          |
| `#_foo\|`  | `#_foo`          |
| `\|#_foo`  | `#_foo`          |
| `#_ \|foo`  | `#_foo`          |
